modular_skyrat/modules/sec_haul/code/melee/baton.dm

```diff
@@ -0,0 +1,556 @@
+	. = ..()
+
+	set_light_range_power_color(default_light_range, default_light_power, default_light_color)
+	if (can_overcharge)
+		add_item_action(/datum/action/item_action/stun_baton/toggle_overcharge)
+
+/obj/item/melee/baton/security/ui_action_click(mob/user, actiontype)
+	if (istype(actiontype, /datum/action/item_action/stun_baton/toggle_overcharge))
+		if (!charging)
+			if (active)
+				toggle_overcharge(user)
+			else
+				balloon_alert(user, "inactive!")
+		else
+			balloon_alert(user, "already charging!")
+		return
+
+	return ..()
+
+/obj/item/melee/baton/security/add_baton_notes()
+	. = ..()
+
+	. += ""
+	. += "Overcharge changes:"
+	. += "Stun armor penetration increased by [span_blue("[BATON_OVERCHARGE_ARMOR_PENETRATION_PERCENT*100]%")] + [span_blue("[BATON_OVERCHARGE_ARMOR_PENETRATION_FLAT]")]"
+	. += "Hits to knockdown reduced by [span_blue("[overcharge_knockdown_swing_reduction]")]"
+	. += ""
+	. += "Stun power usage increased by [span_red(("[overcharge_cell_cost_mult*100]%"))]"
+	. += "Passively discharges [span_red("[overcharge_passive_power_loss]")] volts and [span_red("[overcharge_passive_power_loss_percent*100]%")] of the installed cell's capacity per second"
+	// giving info to people that theyd have to find out through trial and error otherwise is a good thing imo
+
+/**
+ * Switches the overcharge state to the opposite of what it currently is.
+ *
+ * Args:
+ * mob/user: The mob that initiated the toggle.
+ *
+ * Returns:
+ * True if a state change occurs, false otherwise (was charging, no cell, not enough power, etc)
+ */
+/obj/item/melee/baton/security/proc/toggle_overcharge(mob/user)
+
+	if (charging)
+		return FALSE
+
+	if (overcharged)
+		disable_overcharge(user)
+	else
+		if (!cell)
+			if (user)
+				balloon_alert(user, "no cell!")
+			return FALSE
+
+		if (cell.charge < (cell_hit_cost * overcharge_cell_cost_mult))
+			if (user)
+				balloon_alert(user, "not enough power!")
+			return FALSE
+
+		if (user)
+			user.balloon_alert_to_viewers("charging...")
+		else
+			balloon_alert_to_viewers("charging...")
+
+		playsound(src, 'sound/weapons/flash.ogg', 110, TRUE, -1, frequency = 0.7) // very nice "charge-up" sound
+		addtimer(CALLBACK(src, PROC_REF(enable_overcharge), user), overcharge_time)
+		charging = TRUE
+
+	return TRUE
+
+/**
+ * Enables overcharge, a high-power mode that massively increases armor penetration and knockdown ability at the cost of
+ * extreme stun cell cost and passive power discharge.
+ *
+ * Often is called after a delay from toggle_overcharge.
+ * This delay, coupled with the passive discharge, encourages users to only keep overcharge active
+ * when absolutely necessary, and think stratetegically about when to use it.
+ *
+ * Plus, the high cell cost and discharge both incentivise users to upgrade their cell, which is always good.
+ *
+ * Args:
+ * mob/user: The person that initiated the overcharge attempt. Nullable.
+ */
+/obj/item/melee/baton/security/proc/enable_overcharge(mob/user)
```

Comment:
You must add a chunky fingers check so overcharge cannot be enabled while under a condition the stun cannot be used normally.